proposals/lip-introduce-tree-based-key-derivation-and-account-recovery.md

```diff
@@ -0,0 +1,274 @@
+
+
+##### Returns
+
+A BLS secret key.
+
+##### Execution
+
+```python
+getBLSPrivateKeyFromPhraseAndPath(phrase: bip39Mnemonic, path: list[uint32]):
+    masterSeed = bip39.toSeed(phrase)
+    key = derive_master_SK(masterSeed)
+
+    for index in path:
+        key = derive_child_SK(key, index)
+    return key
+```
+
+The logic of the functions `derive_master_SK` and `derive_child_SK` are specified in [EIP 2333](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-2333.md).
+
+
+#### Derivation Path for BLS Keys
+
+Similarly to the specifications of [EIP 2334](https://eips.ethereum.org/EIPS/eip-2334), we use the path `m/12381/134/chainID/0` for the BLS key derived for the chain with ID `chainID`.
+
+If `chainID` is not known, use the path `m/12381/134/0/n` for the n-th generator key derived for chains without known chain ID.
+
+In particular, the path used to derive the BLS private key for the Lisk mainchain is `m/12381/134/1/0`.
+
+
+### Secret Recovery Phrase
+
+The basic entropy required to generate the various keys used in the Lisk ecosystem is stored in a 12 or 24 word phrase, as described by [BIP 39](https://github.com/bitcoin/bips/blob/master/bip-0039.mediawiki).
+
+For users wanting to run a generator node, we recommend using a 24 word secret recovery phrase. This would be closer to specifications from [draft-irtf-cfrg-bls-signature-04](https://datatracker.ietf.org/doc/html/draft-irtf-cfrg-bls-signature-04#section-2.3), even though the BLS key derivation specified here is different from the one specified in the IETF draft.
+
+
+### Account Recovery
+
+When inputting a secret recovery phrase into the Lisk wallet, the wallet will try to assess if this phrase has been used in the past and if it corresponds to any account which at one point received tokens.
+
+Given a secret recovery phrase `phrase`, the potential keys previously used by users are discovered by the following method, setting `GAP_LIMIT = 20`, as in [BIP 44](https://github.com/bitcoin/bips/blob/master/bip-0044.mediawiki#address-gap-limit):
+
+
+1. Let `usedDerivationPath = []`.
+2. Check for an account corresponding to the given `phrase` using the legacy key derivation method (`private Ed25519 key = SHA-256(phrase)`). If the account corresponding to this private key has at least 1 incoming transaction, append `legacy` to `usedDerivationPath`.
+3. Set `n = 0`, `gap = 0`.
+    1. Derive the key pair corresponding to `phrase` and `path=m/44'/134'/n'`. 
+    2. If the corresponding account has at least 1 incoming transaction, append `path` to `usedDerivationPath` and reset `gap=0`.
+    3. If the corresponding account has no incoming transactions, increment `gap`. If `gap>GAP_LIMIT`, stop the account discovery.
+    4. Increase the account index `n` and go back to step `a`.
+4. Return `usedDerivationPath`.
+
```

Comment:
written in code, please check it carefully :) We restrict to account with at least 1 transaction simply because if they do, it means the account exists.